### PR TITLE
ChatLog v3

### PR DESCRIPTION
**Pull request by krepa098**
This is my work on the QGraphicsScene based chat log I've been working on during the last few months. The main goal was to reduce memory consumption and fix all the glitches of the html based chat log (smiley alignment, elided text, ...).

Other benefits:
- Custom text selection 
- Animations
- Widgets inside the chat log
- No need for base64 encoded smileys

Todo:
- [x] Check whether message history is broken or not
- [ ] Possibly change the selection algorithm to what @zetok suggested
- [x] Test with lots of messages
- [x] Implement quoting
- [x] Respect time format
- [x] fix https://gist.github.com/zetok/773876e11737200564c4
- [x] fix "Tox User" in group chats
- [x] fix anchors
- [ ] fix autoscroll (can't reproduce)
- [x] move \* to name col (done, needs color)
- [x] "load history" button
- [x] revisit selection rectangle
- [x] respect "use emoticons"
- [x] clear should reset previousId
- [x] fix crash in combination with history and clear
- [x] implement scroll while selecting
- [x] get rid of html in copied text
- [x] typing notification

I mainly created this PR to get some feedback/bug-reports. This is not ready to be merged although it should be fairly usable and stable.

Pic:
interface3

**EDIT**: New branch at https://github.com/krepa098/qTox/commits/chatlog_v3_1 (unstashed version as requested by @dubslow )

### CODE REVIEW

**Comment by zetok** [comment]
As expected, upon trying to use it I've encountered segfault.

@krepa098's branch with patches (#920, #962, #972) applied

https://gist.github.com/zetok/773876e11737200564c4

**Comment by zetok** [comment]
Without #972 it wasn't reproducible

**Comment by zetok** [comment]
joins/parts

When `Zetok\0` peer left chat, for some reason qTox showed `:3`, as if that was the peer who left. Also, upon `Zetok\0` re-join, there's shown `Tox User` joining, instead of `Zetok\0`

**Comment by zetok** [comment]
Actually it's reproducible without #972, crashes only when clicking on particular contacts

**Comment by zetok** [comment]
feedback

Edit: BTW, new chat form doesn't display time in specified format – i.e., it should display also seconds on my every screenshot, but they're not displayed

**Comment by zetok** [comment]
greentexting doesn't work -.-'
why no greentext

File: src/chatlog/content/text.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    return text;
+}
+
+void Text::ensureIntegrity()
+{
+    if(!doc)
+    {
+        doc = new CustomTextDocument();
+        doc->setUndoRedoEnabled(false);
+        doc->setUseDesignMetrics(true);
```
Comment (apprb) [inline src/chatlog/content/text.cpp]: strange font rendering

png1

Top variant: original
Bottom varian: with commented `doc->setUseDesignMetrics(true);` string

**Comment by zetok** [comment]
Hm, there's no way (or at least I wasn't able to spot it quickly) to load older than 1 week history.

**Comment by dubslow** [comment]
I would very much prefer that merged version doesn't have a squashed history... and the commits added here should preferably make it back to the unsquashed version

(Still don't have a computer at the moment)

**Comment by zetok** [comment]
action messages
action messages don't look good, could `*` be located right under `i`, and nick be distinguishable (bold?) from message part?
On screenshot whole `:3 urationValue of the *active* configura` is a nick, while just `asdf` is a message..

**Comment by ProMcTagonist** [comment]
Could we get the centered boxes back instead of ( i ) messages? They look better, especially when there are a lot of them together.

**Comment by dubslow** [comment]
@zetok How does name compare to the message in the current implementation?

**Comment by zetok** [comment]
@krepa098 after trying to send test text sample:
text test sample
I din't wait long enough to check how much more RAM qTox could actually eat, since initially after pressing "send" it used <200MB, and grew up to ~2GB

What I did:
1. Send FR to contact.
2. Before FR was accepted, I've pasted and "sent" text; qTox "hanged" right at that moment.
3. After receiving FR on other client I've accepted it and messages were received from time to time (usually few messages per minute) from hanged qTox.
**EDIT:** Received messages were not the ones that were supposed to be sent first. I.e., received messages were somewhere  from ~middle(? dunno really, just guessing) of my test sample.

# 

@dubslow it doesn't?
muh name

**Comment by dubslow** [comment]
And you'd like that behaviour to change both in the new implementation and the current one?

The messages appearing out of order is likely an issue with the DB/faux offline messaging, rather than the chat form. @apprb

**Comment by zetok** [comment]
Why bother changing behaviour/look of old chat form that will be deprecated as soon as new one stops crashing qTox and will get missing features?

# 

Edit: not sure if I included it in my previous comments if not;
There are some problems with autoscrolling in chat; not sure what triggers it, but chat isn't always automatically scrolled down like it was in old chat form.

**Comment by zetok** [comment]
no, bad link, bad!
original text:

```
1. Apply patch https://gist.github.com/zetok/a326be414c73eb64c48f
2. Recompile core and client
```

Reproducible: always

**EDIT**: another sample
muh link
Not a real issue, just doens't look good

**Comment by zetok** [comment]
distance
↑ ~~there's quite a lot of distance between those messages, I don't know why.~~ Actually ↓
missing msg :c

Also, I've noticed that once nick of person who has sent messages wasn't displayed in gropuchat history. After selecting text name of that peer appeared. It didn't contain any special characters AFAIK.

In the same way as nicks, some messages in groupchat history are missing.

By `groupchat history` I mean that I had to scroll way back to see them.
no wai
↑ wrong, same part of chat after selecting message ↓
oh noes

**Comment by zetok** [comment]
too muh select
selection not always selects whole nick
qTox e8ca000379267aaf7e60c8f20e6adcbecaedd2fd with previously mentioned patches applied

**Comment by zetok** [comment]
emote setting
↑ emoticons shouldn't be displayed according to settings… ↓
nope emote

**Comment by zetok** [comment]
This PR fixes #906, but at the same time invalidates #857

**Comment by zetok** [comment]
After building patched qTox based on branch chatlog_v3_1 a0a56e32a0fcf333dbf720d46dbf4378a352e7fa so far I can no longer reproduce segfault

**Comment by zetok** [comment]
**pause button** doesn't seem to work in file transfer widget.

Also, on a0a56e32a0fcf333dbf720d46dbf4378a352e7fa it's gotten worse with **displayed nicks**
nick, no nicks

# 

also there's bug with how nick is displayed when it's long
long nick

**Comment by apprb** [comment]
related to a0a56e3
In `GenericChatForm::addAlertMessage` and `GenericChatForm::addMessage`
`resolveToxID(..)` function should be used instead of `Core::getInstance()->getPeerName(..)` in order to get nick of every speaker in a groupchat, not only ones in your friendlist.

**Comment by zetok** [comment]
Copying multiple messages in groupchat doesn't work well if one has a null-terminated nick and tries to copy message. Basically everything after NUL is not copied.

**Comment by dubslow** [comment]
Gee... what a shocker... maybe we should truncate that in the UI... and thus in the chatform... well obviously we should, I wish I could get around to it

**Comment by zetok** [comment]
on a74b9ce11cea45f3ad62a0e16ede8576fb12425d loading history doesn't work

**Comment by zetok** [comment]
no idea whether latest commits were supposed to fix history loading, but it doesn't seem to work with older history on 6711fd6ee4a6fd2c40db4f89dff040b81025e3ec

What I'm doing to load all history: load history since 2013.

I have sqlite db full of messages, all weights 11.5 MB, just for testing how well new chatform is supposed to handle lots of messages.

**Comment by krepa098** [comment]
Does it freeze? Does it work with the current chat log?

**Comment by zetok** [comment]
~~Doesn't freeze, just nothing happens
I have 11.5 MB big db, from which apparently I can't load history, no matter whether using new or old chat form..~~
I've found where problem lies. For some reason, qTox cannot into db loading.
How to reproduce:
1. Get some db with messages older than 1 week
2. Click load messages
3. Set year to before 2014
4. Load ← history is not being loaded.

↑ if loading history by going just by few months, then it works. 
@apprb ?

@krepa098 and yes, when I finally managed to load db, qTox "hanged". I guess it tries to load those 11.5 MB of messages, by using 100% of CPU and very slowly growing in RAM usage. Other client shows "hanged" qTox as online.

**Comment by zetok** [comment]
So, since there's a performance issue with loading history, some perf with flamegraph: https://a.pomf.se/gtlksz.svg

**Comment by suhr** [comment]
@krepa098 Does it do layout for every new line? And does it layout everything or some part only?

**Comment by zetok** [comment]
new
↑ new chat form

↓ old chat form
whoever s

and in µTox: https://img.bi/#/XeX5Imo!KN0tzUNgdpaqAoECLlsCv94RRhLFk9nofBEauVH4

I have no idea which way is correct.

**Comment by apprb** [comment]
@zetok can't reproduce your history loading issue

@krepa098 
Every message loaded from history marked as sent even it isn't true.
Patch to fix it:
https://gist.github.com/apprb/e4281007d924492c6731

**Comment by zetok** [comment]
@apprb Right. 
Between steps 

```
3. Set year to before 2014
4. Load
```

don't click on day, just set year (I did it by clicking on year and using mouse scroll) and click load.

Loading seemed to work when clicked on day.

**Comment by apprb** [comment]
Hm, you right. Seems like it is how `QCalendarWidget` works, such actions just show corresponding month's layout.

UPDATE: fix proposed in #983

**Comment by zetok** [comment]
I can't believe it. Amount of messages that with older chatform were causing OOM killer to be triggered now doesn't look like a problem, at least from memory standpoint on 4f01f5ec0e11db2f7c9b541fe13be09fe5dab0de

So, before loading history qTox used ~127 MB of RAM. After whole history for contact was loaded, RAM usage increased to 234 MB (tried 2 times, consistent result). Earlier same messages would eat >12GB(!) of RAM.

There are still some minor (in comparison to what was earlier) performance problems with loading history, during time it loads qTox remains hanged. Now though it takes only ~2-3 minutes to load whole history, whereas a bit earlier after >40 min history still didn't seem to be loaded.

Flamegraph of loading history: http://a.pomf.se/rwmkwu.svg

After history is loaded, there are some performance problems when scrolling through history (qTox for short moments hangs).

I haven't made flamegraph that would include performance problems with scrolling, since I have no idea to what degree it would be useful, considering that loading history itself is quite heavy task.

If such flamegraph would be useful, just say so.

Still, improvement is tremendous, thanks @krepa098, it's really awesome :3

**Comment by krepa098** [comment]
> If such flamegraph would be useful, just say so.

Yes, please.

> After history is loaded, there are some performance problems when scrolling through history (qTox for short moments hangs).

There is a trade-off between memory consumption and compute time.

**Comment by zetok** [comment]
Flamegraph for scrolling after loading history: https://a.pomf.se/ihovbl.svg

**Comment by zetok** [comment]
nicks
↑ that's on version which had nicks fixed in groupchat, but it's not the latest commit; for some reason @subliun's nick after action messages wasn't displayed (I'm guessing that it's his nick)

**Comment by krepa098** [comment]
> Flamegraph for scrolling after loading history: https://a.pomf.se/ihovbl.svg

Thanks. I probably got rid of the Style::resolve part.

**Comment by zetok** [comment]
Scrolling flamegraph on af575c09e6612c93c3153e1dca64704bcf1c5563: https://a.pomf.se/jlstnv.svg
Faster, but <1s "stuttering" is still present when scrolling by mouse scroll, and short hangs when using scroll bar.

**Comment by zetok** [comment]
b92d4f3d2f275b8c83765c9f1c5adca3df8c96d1
I've loaded history, scrolled for a bit and tried selecting text/messages from ~middle of history.

Selecting text is really slow. It takes several seconds to select some 'line' text.
Selecting few messages is also slow.
Deselecting(!) selected 'line' text is also really slow.
Deselecting selected messages if only a bit slow, much faster than deselecting 'line' text.

Flamegraph: https://a.pomf.se/egfmbr.svg

Edit: I think that scrolling is a bit faster than on earlier commits, but since I'm not sure to what degree placebo effect may affect me…

Edit2: All those things have been tested/measured on qTox compiled with `-O0`. Since most likely qTox will be ran almost always compiled with `-O2`, would it make significant difference? I.e., should I test its performance and provide flamegrap with `-O2`?

**Comment by krepa098** [comment]
Weird... BspTreeIndex is supposed to speed up operations like itemAt() but it doesn't seem to work. Could you please test 6c6d0254fc32583f2802847adcc5579d14bb5943 to see whether or not it's faster than 6c6d0254fc32583f2802847adcc5579d14bb5943?

> Edit2: All those things have been tested/measured on qTox compiled with -O0. Since most likely qTox will be ran almost always compiled with -O2, would it make significant difference? I.e., should I test its performance and provide flamegrap with -O2?

I don't know. Yes if profiling works fine with O2 enabled.

**Comment by zetok** [comment]
I've checked out 6c6d0254fc32583f2802847adcc5579d14bb5943 and it seems to be faster than b92d4f3d2f275b8c83765c9f1c5adca3df8c96d1.

**Comment by zetok** [comment]
1bf08981212bc5c265669ed9efb5b77669177926 with `-O2`
Even though it's optimized, it still seemed to be slower than 6c6d0254fc32583f2802847adcc5579d14bb5943 when scrolling.

Flamegraph: https://a.pomf.se/xqwwfz.svg

**Comment by zetok** [comment]
Problem with autoscrolling seems to appear most often when I reply after I've received message in 1 on 1 chat, or the other way around. Perhaps special symbol in my nick somehow affects it?

Also it's noticeable that after receiving message autoscroll doesn't scroll fully, leaving some bit of "space" unscrolled

**Comment by zetok** [comment]
on a0693c2fc9135d7f76aa461f4847844cc0577a63 I'm unable to save files from incoming file transfers
Edit: there's shown error window about not sufficient permissions to save file, even though I have sufficient permissions to do so..

Edit2: I think that there is a need for animated typing notification

**Comment by krepa098** [comment]
What kind of animation? Like Skype? I don't think that's entirely possibly as we get no backspace/typing events.

EDIT: Please don't edit your post if you have something important to say as I don't get notified via e-mail :)

**Comment by zetok** [comment]
I don't know what kind of typing notification skype has, but I think that simple 3 dots appearing one after another in place of underscore would be nice (some other idea for animation?).

I don't mean that animation should _actually_ show whether someone is writing or not, but rather to show for user that qTox "works" and didn't just randomly freeze, since e.g. for me every time I see unmoving typing notification image I have a feeling that qTox just froze..

**Comment by zetok** [comment]
typing notification & status message
Typing notification is kinda interfering with status changes

Status change:
Online → Away → Online

on b4fa8323af5e8834812e1a565067f1e87a4a7e90

**Comment by krepa098** [comment]
>  I think that simple 3 dots appearing one after another in place of underscore would be nice

Probably a good idea. I'll have to get rid of the stylus though... 16px isn't really "a lot of space".

> Typing notification is kinda interfering with status changes

Yes, I forgot to reposition the notification.

**Comment by zetok** [comment]
qTox b4fa8323af5e8834812e1a565067f1e87a4a7e90 segfaulted
https://gist.github.com/zetok/7e2fcd81f84c2a6df62e

I doubt that it's really reproducible, since it happened when I was selecting text (multiple messages), and it didn't segfault ever before when doing that (on a same version too).

Perhaps when I was selecting text there appeared new message in groupchat, could that matter?

**Comment by zetok** [comment]
I think that spinning thing doesn't really look good. It's just my opinion, but even that unmoving notification was better.
https://a.pomf.se/usgpek.webm

**Comment by ProMcTagonist** [comment]
> https://a.pomf.se/usgpek.webm

It's not good. It seems like it has to do with status changing because it's made out of status orbs. Completely arbitrary and confusing. If you're going to have a pretty graphic for typing indications, it should be like the one in iOS or Facebook chat: a speech bubble with an animated ellipsis.

**Comment by ghost** [comment]
Too flashy and useless. Why not just put "xyz is typing" or ye olde three flashing dots above the input text field? Also that spinner looks off center and is bound to kick off someones OCD.

**Comment by suhr** [comment]
Why not use `. -> .. -> ... -> .` animation?

**Comment by krepa098** [comment]
I'd like to have something fluid... I do have an idea but I don't know how to implement it yet.
(No, it's not about spinning dots! :wink:  ).

**Comment by zetok** [comment]
Hm, on 5429e50bd0c4454c2158d14014a0f87e27442208 and on some earlier commits I've encountered bug, which causes parts of messages, or whole messages to not be displayed.

I'm not exactly sure what causes it, since it's rare, but in occurrences where it happened, it seemed to have something to do with me selecting messages and scrolling(?) by selection.

I've "scrolled" with selection up and down without releasing selection.

**Comment by zetok** [comment]
missing message

Edit: makes you wonder why that message wasn't displayed :D

**Comment by krepa098** [comment]
Fixed by 2c8e7c638d06fb464b0e3985fe46f1a5698836c3 ?

**Comment by zetok** [comment]
on 5429e50bd0c4454c2158d14014a0f87e27442208 
selection
I don't really know how to reproduce, since it rarely happens.

# 

I haven't noticed "missing" messages on newer version, but I didn't use it for long…

**Comment by krepa098** [comment]
> I don't really know how to reproduce, since it rarely happens.

Don't worry. I know why it happens.

**Comment by zetok** [comment]
Log saved from r-click doesn't have timestamps of any kind, I think that's a major setback.
When saving log, could it be saved in format
`[timestamp in hh:mm:ss] nick`
?

**Comment by zetok** [comment]
topic message
notification about setting topic message is borked with certain content

thanks to @subliun

**Comment by zetok** [comment]
can 882d19d311ee6bf85d252181506fbb5880571138 be reverted? `…` isn't really informative..

# 

on cfe4458d37d7efac51046d88ff3ba41e49c1b8ed:
no typing notification
I wrote to offline `:3` from  `:3`

Not sure if that's the way it's supposed to be, there's no faux offline messaging notification nor date.

Additionally to that, there's **huge** increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg

**Comment by krepa098** [comment]
> can 882d19d be reverted? … isn't really informative..

Steam does it too. The "... is typing" part seemed rather redundant to me.

> I wrote to offline :3 from :3

You're not supposed to be able to add yourself :). 

> Not sure if that's the way it's supposed to be, there's no faux offline messaging notification nor date.

Not sure about this... does it also happen using the current chatlog?

> Additionally to that, there's huge increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg

This only applies to the case above, right?

**Comment by zetok** [comment]
> > I wrote to offline :3 from :3
> 
> You're not supposed to be able to add yourself :). 

um, no, those are 2 separate profiles/IDs, just use same name

> > Additionally to that, there's huge increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg
> 
> This only applies to the case above, right?

~Yes. Beside that, I think that there's recently been some slight performance regression to chatform_3_1 with loading history from DB (takes more time compared to what I remember). I'll see about making flamegraph of it when I'll have time.

> Not sure about this... does it also happen using the current chatlog?

Sadly, amount of messages I load causes old chatform to be killed by OOM killer after consuming 12GB of memory, so I'm unable to test it with it…

**Comment by zetok** [comment]
missing notification images

Notification about topic change is missing an indicator

**Comment by zetok** [comment]
copying from groupchat doesn't work. no matter to which clipboard on Linux; seems like regression of some kind

**Comment by krepa098** [comment]
> Notification about topic change is missing an indicator
> copying from groupchat doesn't work. no matter to which clipboard on Linux; seems like regression of some kind

bildschirmfoto4

Seems to work fine for me. Did you try a fresh clone, make?

> Additionally to that, there's huge increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg

This should be fixed now.

**Comment by zetok** [comment]
> Seems to work fine for me. Did you try a fresh clone, make?

Portage makes sure that clone is fresh, and always makes a clean build.

Issue seems to be reproducible on 783caf932c66e685c5053bef28fbf9c1baae3bc6

**Comment by dubslow** [comment]
@krepa098 Would you be willing to force push the other branch to here? It would make following the conversation rather easier lol

**Comment by zetok** [comment]
on 39b2771b8dfb70c3e7b5438f8c00285b60f93ab8

muh link?

**Comment by zetok** [comment]
issue with word wrapping affects all text (e.g. topic changes), not just messages

**Comment by zetok** [comment]
some strange crash when closing qTox: https://gist.github.com/zetok/7ba7baa413cbbe5ca5a9

**Comment by zetok** [comment]
on ~~39b2771b8dfb70c3e7b5438f8c00285b60f93ab8~~ EDIT: it's actually on 09291bad52fb84fd716638fccdfda5a429562589 , I've copypasted wrong hash

Pending file transfer:
button pls
- no buttons
- speed shouldn't be displayed
- no need for progress bar(?)

selection pls

**Comment by zetok** [comment]
highlight

**Comment by krepa098** [comment]
Seems like the SVGs don't show up. Make sure you've got qt5-svg installed (or whatever it's called on gentoo) and retry on 038a7fc.

**Comment by zetok** [comment]
oh, so this PR introduces dependency on qtsvg to work properly? If so, then it should be reflected in `INSTALL.md` .
After installing qtsvg:5 ~~it seems to work properly~~ EDIT: icons show up.

**Comment by zetok** [comment]
wait for it..
dun
pls accept

Looks nice (?)
And animation when file transfer finishes is awesome.

**Comment by suhr** [comment]
It's somewhat out of canon.
cast @ItsDuke

**Comment by krepa098** [comment]
> oh, so this PR introduces dependency on qtsvg to work properly? 

Ya, I think SVGs are the best way to tackle HiDPI support as they get rid of these `_x2.png` files.

**Comment by calimeroteknik** [comment]
Round buttons are inconsistent with the style of the message box, so I suggest giving up on these.
qtoxmsg

Also, the spinners don't add functionality, there already was a red mark. Is it really necessary to grab the user's attention? That's what animated stuff does.

**Comment by zetok** [comment]
@calimeroteknik yes, spinners are necessary. They're the best: https://a.pomf.se/xanbyd.webm

**Comment by dubslow** [comment]
Can the spinner appear next to, but not in place of, the timestamp? Just testing this for the first time since there's only one major outstanding branch now :)

**Comment by ProMcTagonist** [comment]
Is the timestamp the time the message was created, the time it was successfully sent, or the time it was received?

**Comment by calimeroteknik** [comment]
I tested this out earlier (5bd3cd0), and the answer was "the time it was created".

**Comment by zetok** [comment]
on 5e4e56778ba93e2048925a5b21e3d7a694b6ad4f
There's an unnecessary pop-up window about not sufficient permissions to save file.

Steps to reproduce:
1. "send" file to qTox
2. click to save file in qTox
3. in file picker click "cancel"
4. unnecessary warning about permissions shows up

**Comment by zetok** [comment]
on 09291bad52fb84fd716638fccdfda5a429562589
duplicated message

For some reason messages were displayed 2 times.
(don't worry about missing .svg, it's an instance since before I've installed qtsvg:5 )

**Comment by krepa098** [comment]
> Is the timestamp the time the message was created, the time it was successfully sent, or the time it was received?

It _should_ be the time they were successfully sent.

> Can the spinner appear next to, but not in place of, the timestamp?

Well, no (technically, yes :p). Imo it doesn't make sense to show a timestamp on an undelivered message. If you sent a message on 11:00 and your friend received it on 11:05 you probably want it to display 11:05.

> For some reason messages were displayed 2 times.

Related to #1089?

> Also, the spinners don't add functionality

Sure. They tell you "it's being worked on" whereas a red line reads more like an error. I'll make them turn slower if that helps?

EDIT:

> I tested this out earlier (5bd3cd0), and the answer was "the time it was created".

Works fine for me?
bildschirmfoto6

**Comment by zetok** [comment]
@krepa098 no, spinners were perfect. There's no need to change them (haven't seen slower spinners yet)

**Comment by zetok** [comment]
> > For some reason messages were displayed 2 times.
> 
> Related to #1089?

No. Look at timestamps

**Comment by calimeroteknik** [comment]
> They [spinners] tell you "it's being worked on" whereas a red line reads more like an error. I'll make them turn slower if that helps?

I finally got your idea! And so I have a suggestion.
Usually, a (fast) spinner means "I'm working on it but the machine/network is slow".
It's not the case here, I think? More like "idly waiting for your contact to be online".

So yeah, I believe that making them slower is good.
I tried to look for alternatives but have nothing else to suggest, sorry. And zetok likes spinners… :)

**Comment by zetok** [comment]
@calimeroteknik I'm sorry to say this, but please stop with this stupid bullshit about slow spinners. Now they look ugly, retarded, and qTox's "speed" decreased dramatically, to the point where it no longer feels "cool" to use it.

> Usually, a (fast) spinner means "I'm working on it but the machine/network is slow".

Bullshit. Fast spinner means that your machine is fast.

If all other programs have fast spinners (chrome, ff), and qTox doesn't have them, it looks simply as if it was slowing whole system down, thus creating an impression of bloated software that cannot even deliver messages properly.

After all, everyone knows that the faster spinners spin, the faster is your box.

EDIT:
https://a.pomf.se/nccnbw.webm _argh, my eyez!_

**Comment by suhr** [comment]
6.5 and 6.6 are really good values for speed. Consider using 6.6 for the spinners speed.

**Comment by suhr** [comment]
Ctrl+Insert doesn't work for copying text, neither works select to copy & middle click to paste (often used feature in linux).

**Comment by dubslow** [comment]
I rather disagree with a slow spinner, I agree with what zetok said.

**Comment by compliant** [comment]
Actually, a spinner is not adapted for indicating a pending message.

A spinner indicates that a background action is currently being performed ([see wikipedia](http://en.wikipedia.org/wiki/Throbber)). There is no action performed when a message is pending: the application is idle, waiting for the recipient to connect.

A better indicator would be a static _status_ indicator. Some examples:
- the red bar (change its color if it seems to mean "error");
- a static icon near the messages (a colored dot, an arrow, an elipsis "…", or a clock for "delayed");
- maybe an horizontal bar at the last position the receiver was seen online, all messages under this bar being "pending".

**Comment by suhr** [comment]
mb different indication should be used for “user looks online, but message is not delivered yet” and “user is offline and message will be sended later”?

**Comment by calimeroteknik** [comment]
Very true. A spinner would definitely be adequate when trying to deliver the message to an online user, but when the message is stored to be sent later, some static icon is preferrable. (sandglass, clock, suspension dots, offline user?)

**Comment by krepa098** [comment]
> mb different indication should be used for “user looks online, but message is not delivered yet” and “user is offline and message will be sended later”?
> 
> Very true. A spinner would definitely be adequate when trying to deliver the message to an online user, but when the message is stored to be sent later, some static icon is preferrable

Yes, that's pretty reasonable.

> sandglass, clock, suspension dots, offline user?

... envelope?

Anyway, I'm not going to add this now as it's not really that important and I preferably want this to get merged sooner than later.

> No. Look at timestamps

@zetok did you manage to reproduce it?

**Comment by zetok** [comment]
@krepa098 no, after qTox's restart it was shown only once, as it should be

**Comment by dubslow** [comment]
I also would like to get this merged as quickly as possible, even with some bugs remaining, because it's a major improvement over the current state of things.

On the other hand, discussion here has mostly devolved to nitpicks, rather than major issues that render merging a bad idea at the moment.

@krepa098 Can you enable issues in your repository, so that we may get a clear listing of the deal-breaking problems? Once the major problems are fixed I would definitely love to merge this.

**Comment by zetok** [comment]
Well, as far as I am concerned, chatlog v3 is in a shape to be merged, [few remaining issues](https://github.com/krepa098/qTox/issues) don't look like merge blocker, and are something that can be ~easily (am I right?) fixed later on.

(tl;dr merge! - apparently that segfault is master tux3/qTox problem)

**Comment by dubslow** [comment]
@tux3 @krepa098 

I'd like to pin down the segfault before merging, to be sure that it is actually a problem on master (and not on this). That'll have to wait until tomorrow though, giving time for the others to weigh in.

**Comment by zetok** [comment]
Hm, I've encountered performance problem when sending **large** amounts of text (splitting); qTox is not responding, uses 100% CPU and mem usage increased from ~200 MB to >600MB before I've killed it (didn't increase rapidly). Not a merge blocker IMHO (considering amount of text), but…

Flamegraph: https://a.pomf.se/toktav.svg

**Comment by suhr** [comment]
FIX: Underline links only on hover, change cursor on them.

**Comment by ItsDuke** [comment]
@suhr @zetok sorry about being a little late with this, but you're correct. For boxes like that, it's best to refer back to the original mockup:

capture

Obviously I don't know how close you'd like to get to it, but just thought I'd post it just to give an idea.

**Comment by calimeroteknik** [comment]
+1 Yes, I pointed out earlier that the change on the buttons of the download box was inconsistent with the rest of the UI (typing field), and seems to differ from he UI guidelines.
It would be best to revert to the appearance of the above screenshot.

**Comment by ProMcTagonist** [comment]
+1 @ItsDuke

**Comment by dubslow** [comment]
@krepa098 Could you please restore the file transfer widget to match the mockup, as requested?

**Comment by krepa098** [comment]
Sure. May take some time, though.

**Comment by suhr** [comment]
BUG: cropped _posted_ emoji in new chatform when you change emoticon size from small to big
How to reproduce:
- Post an emoticon in the chat
- Change emoticon size in the settings to a bigger one
- Return to the chat

**Comment by suhr** [comment]
Triple click doesn't select the whole line.

**Comment by krepa098** [comment]
bildschirmfoto9

That's how it looks now. I'll remove the 0kiB/s from paused and unaccepted file transfers.

**Comment by calimeroteknik** [comment]
@krepa098 nice!
Hum, why 1 pixel of spacing instead of 2? The textbox to write messages has 2px of spacing so why not do the same?

I have an issue with the text color (dark Qt theme), on this pull request's branch:
qtoxwhite
On a broader debate, I'm not very happy about qTox's white background with my dark theme, but well.
At least keeping the contrast intact would be necessary!

Lastly, we reached earlier the conclusion of a static icon, for messages awaiting the other person to connect. You said an envelope or something, maybe something like mail-pending ? [svg](http://sprunge.us/KiKX) (just a mockup)

**Comment by suhr** [comment]
> dark theme

Well, time to create an issue: https://github.com/ItsDuke/Tox-UI/issues/12

**Comment by dubslow** [comment]
Major problem. The merge has been force pushed back. It increases the Qt requirements to 5.3 (QMarginsF), and that's a massive massive change that I'm not entirely sure the build system can handle. It was already a huge pain in the butt to get 5.2 working...

@krepa098 Is there an easy way for you to reduce the requirements to 5.2?

**Comment by ProMcTagonist** [comment]
What state causes the yellow buttons? @krepa098 

@calimeroteknik tacking an icon onto every line that's awaiting sending sounds really cluttered and unnecessary. I'm in favor of the red bar, maybe with color adjustments.

**Comment by calimeroteknik** [comment]
@ProMcTagonist currently (in this pull request) it's spinners, which is even worse for pending messages. I was also happy with the current red bar.

**Comment by dubslow** [comment]
I like the moving icon, and I'm reasonably certain zetok does as well.

**Comment by krepa098** [comment]
> Major problem. The merge has been force pushed back. It increases the Qt requirements to 5.3 (QMarginsF), and that's a massive massive change that I'm not entirely sure the build system can handle. It was already a huge pain in the butt to get 5.2 working...
> Is there an easy way for you to reduce the requirements to 5.2?

I can use QMargins instead. I wasn't aware of QMarginsF being part of Qt 5.3. But keep in mind that HiDPI support requires a bump to 5.4 as well.

> What state causes the yellow buttons?

It's a file transfer that hasn't been accepted, yet.

> Hum, why 1 pixel of spacing instead of 2? The textbox to write messages has 2px of spacing so why not do the same?

It seems like the spacing is only 1px in the mockup. So I guess the textbox is wrong :P

> Lastly, we reached earlier the conclusion of a static icon, for messages awaiting the other person to connect. You said an envelope or something, maybe something like mail-pending ? svg (just a mockup)

Yes, I remember that. It's on my TODO list but it's not going to be part of this PR. (I like the mockup btw)

**Comment by zetok** [comment]
@krepa098 please don't change faux offline message indicator to something static. Current one is really good.

**Comment by suhr** [comment]
> different indication should be used for “user looks online, but message is not delivered yet” and “user is offline and message will be sended later”

Just use spinners only for the first.

**Comment by dubslow** [comment]
When Debian Jessie becomes stable, we can reexamine Qt versions. (There may still be issues on the Fedora family of things, but Wheezy is a pain in the butt and I don't want to change anything until we can retire Wheezy.)

Edit: If you can, I'd like to get this down to 5.2 and merged as soon as possible -- if you're around, an hour or less.

**Comment by dubslow** [comment]
@krepa098 Since you said they're replaceable, I took a whack at fixing it myself, and Jenkins seemed to compile okay. Having said that, can you look over the commits in master to be sure?

**Comment by krepa098** [comment]
@dubslow I think a did all the changes I wanted to do so I think you can attempt another merge :)

**Comment by krepa098** [comment]
@dubslow Yes that's fine. But there are a bunch of fixes left in my branch (hopefully without regressions).

**Comment by calimeroteknik** [comment]
@krepa098 I just saw that in unicode there are `📨 📩 ✉ 🕒`, this may be of use.
As ProMcTagonist noted, it looks cluttered to have several, since messages will get delivered in order anyway, so one may suffice. On the other hand, with a lot of messages pending, it might be out of sight.
The original red line was indeed nice for pending messages… (spinners being totally OK for messages not yet ACKed)

What could be done about the font color being very like the background, when using a dark theme?
Maybe the fix can come from Tox accepting a darker background for the message pane, but in the meantime it's really unreadable (see earlier screenshot).

**Comment by ProMcTagonist** [comment]
I like spinners for pending, red bar for offline friend (with a firstrun/tooltip message about what's happening).

One thing - spinners should **not** appear immediately. It makes it seem really laggy when every single message has a brief pending period. Maybe only do the spinner after a second.

**Comment by calimeroteknik** [comment]
> Maybe only do the spinner after a second.

Did you try this out? It already fades in with a delay.
Anyway it's cool, since it tells you when the message was acknowledged by your contact's client.

**Comment by ProMcTagonist** [comment]
Looks like I was just testing on a slow bot. It's better everywhere else.

**Pull request merged**